Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+                var bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: selectedEventLoop, host: request.host, port: request.port, configuration: self.configuration) { _ in
+                    self.loopGroup.next().makeSucceededFuture(())
+                }
+
+                if let timeout = resolve(timeout: self.configuration.timeout.connect, deadline: deadline) {
+                    bootstrap = bootstrap.connectTimeout(timeout)
+                }
+
+                let address = HTTPClient.resolveAddress(host: request.host, port: request.port, proxy: self.configuration.proxy)
+
+                // Determe if HTTP/1.1 or HTTP/2 should be used. This is done outside
+                // of the usual client bootstrap `.channelInitializer` because it cannot
+                // be called multiple times like it is currently the case due to Happy Eyeballs
+                return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
+                    let http1PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { pipeline in
+                        pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes).map { _ in
+                            let http1Provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel, parentPool: self), parentPool: self))
+                            protocolChoosenPromise.succeed(http1Provider)
+                        }
+                    }
+
+                    let h2PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { _ in
+                        channel.configureHTTP2Pipeline(mode: .client).map { multiplexer in
+                            let http2Provider = ConnectionProvider.http2(HTTP2ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, channel: channel, multiplexer: multiplexer, parentPool: self))
+                            protocolChoosenPromise.succeed(http2Provider)
+                        }
+                    }
+
+                    if request.useTLS {
+                        _ = channel.pipeline.addSSLHandlerIfNeeded(for: request, tlsConfiguration: self.configuration.tlsConfiguration).flatMap { _ in
+                            channel.pipeline.configureHTTP2SecureUpgrade(h2PipelineConfigurator: h2PipelineConfigurator, http1PipelineConfigurator: http1PipelineConfigurator)
+                        }
+                    } else {
+                        _ = http1PipelineConfigurator(channel.pipeline)
+                    }
+
+                    return protocolChoosenPromise.futureResult.flatMap { provider in
+                        self._connectionProviders[key] = provider
+                        return provider.getConnection(preference: preference)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This function really needs to be broken up into some helper functions.

Reply:
I think there have been a few changes since your review, please let me know if you think this can still be improved 🙂